pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisStreamMetadataProvider.java

```diff
@@ -0,0 +1,177 @@
+    Set<String> shardsEnded = new HashSet<>();
+
+    // TODO: Once we start supporting multiple shards in a PartitionGroup,
+    //  we need to iterate over all shards to check if any of them have reached end
+
+    // Process existing shards. Add them to new list if still consuming from them
+    for (PartitionGroupConsumptionStatus currentPartitionGroupConsumptionStatus : partitionGroupConsumptionStatuses) {
+      KinesisPartitionGroupOffset kinesisStartCheckpoint =
+          (KinesisPartitionGroupOffset) currentPartitionGroupConsumptionStatus.getStartOffset();
+      String shardId = kinesisStartCheckpoint.getShardToStartSequenceMap().keySet().iterator().next();
+      Shard shard = shardIdToShardMap.get(shardId);
+      shardsInCurrent.add(shardId);
+
+      StreamPartitionMsgOffset newStartOffset;
+      StreamPartitionMsgOffset currentEndOffset = currentPartitionGroupConsumptionStatus.getEndOffset();
+      if (currentEndOffset != null) { // Segment DONE (committing/committed)
+        String endingSequenceNumber = shard.sequenceNumberRange().endingSequenceNumber();
+        if (endingSequenceNumber != null) { // Shard has ended, check if we're also done consuming it
+          if (consumedEndOfShard(currentEndOffset, currentPartitionGroupConsumptionStatus)) {
+            shardsEnded.add(shardId);
+            continue; // Shard ended and we're done consuming it. Skip
+          }
+        }
+        newStartOffset = currentEndOffset;
+      } else { // Segment IN_PROGRESS
+        newStartOffset = currentPartitionGroupConsumptionStatus.getStartOffset();
+      }
+      newPartitionGroupMetadatas.add(
+          new PartitionGroupMetadata(currentPartitionGroupConsumptionStatus.getPartitionGroupId(), newStartOffset));
+    }
+
+    // Add new shards. Parent should be null (new table case, very first shards)
+    // OR it should be flagged as reached EOL and completely consumed.
+    for (Map.Entry<String, Shard> entry : shardIdToShardMap.entrySet()) {
+      String newShardId = entry.getKey();
+      if (shardsInCurrent.contains(newShardId)) {
+        continue;
+      }
+      StreamPartitionMsgOffset newStartOffset;
+      Shard newShard = entry.getValue();
+      String parentShardId = newShard.parentShardId();
+
+      if (parentShardId == null || shardsEnded.contains(parentShardId)) {
+        Map<String, String> shardToSequenceNumberMap = new HashMap<>();
+        shardToSequenceNumberMap.put(newShardId, newShard.sequenceNumberRange().startingSequenceNumber());
+        newStartOffset = new KinesisPartitionGroupOffset(shardToSequenceNumberMap);
+        int partitionGroupId = getPartitionGroupIdFromShardId(newShardId);
+        newPartitionGroupMetadatas.add(new PartitionGroupMetadata(partitionGroupId, newStartOffset));
+      }
+    }
+    return newPartitionGroupMetadatas;
+  }
+
+  /**
+   * Converts a shardId string to a partitionGroupId integer by parsing the digits of the shardId
+   * e.g. "shardId-000000000001" becomes 1
```

Comment:
Is shardId guaranteed to be like the above format?  It looks that we probably can assume the convention but the official documentation does not guarantee this specific format. 

https://docs.aws.amazon.com/kinesis/latest/APIReference/API_Shard.html

The safer approach can be keeping the mapping from shardId to the partition group Id not to depend on the conversion of shardId returned by Kinesis.

Reply:
I agree with you. There's no guarantee, but I think it's safe to assume it, because practically, it's always like this. I spoke to few Kinesis users too.
Added a FIXME, to  call this out and re-evaluate this in the future if needed.